File: android/app/src/main/kotlin/com/TheVoteApp/data/gateway/GatewayResolver.kt
package com.TheVoteApp.data.gateway

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.first
import org.json.JSONObject
import java.net.URL
import javax.inject.Inject
import javax.inject.Singleton

private val Context.gatewayDataStore by preferencesDataStore("gateway_cache")

/**
 * GatewayResolver — discovers the live backend API URL without central DNS.
 *
 * Resolution order:
 *   1. Cached URL from a previous successful resolution (DataStore)
 *   2. ENS text record on thevoteapp.eth (via public ENS DoH resolver)
 *   3. Hardcoded fallback gateway CID list in app resources
 *
 * The resolved URL is validated against the known publisher address
 * embedded in the app before use.
 */
@Singleton
class GatewayResolver @Inject constructor(
    @ApplicationContext private val context: Context
) {
    companion object {
        private val CACHED_API_URL = stringPreferencesKey("cached_api_url")

        // Known IPFS gateways to try when resolving the manifest CID
        private val IPFS_GATEWAYS = listOf(
            "https://ipfs.io/ipfs/",
            "https://cloudflare-ipfs.com/ipfs/",
            "https://gateway.pinata.cloud/ipfs/"
        )

        // ENS resolver endpoint (DoH-based, no native ENS library needed)
        private const val ENS_RESOLVER_URL =
            "https://cloudflare-eth.com/v1/mainnet"

        // Known publisher address — embedded at build time
        private const val KNOWN_PUBLISHER_ADDRESS =
            "0x0000000000000000000000000000000000000000" // replaced at release build
    }

    /**
     * Resolve the current API base URL.
     * Falls back through the chain; caches successful results.
     */
    suspend fun resolveApiUrl(): String {
        // 1. Try cache
        val cached = context.gatewayDataStore.data.first()[CACHED_API_URL]
        if (!cached.isNullOrBlank()) return cached

        // 2. Try ENS + IPFS manifest
        val fromEns = tryResolveViaENS()
        if (fromEns != null) {
            cacheUrl(fromEns)
            return fromEns
        }

        // 3. Hardcoded fallback
        return getFallbackUrl()
    }

    /** Call after a successful API response to cache the working URL. */
    suspend fun cacheUrl(url: String) {
        context.gatewayDataStore.edit { it[CACHED_API_URL] = url }
    }

    /** Clear the cache (e.g., after repeated failures force re-resolution). */
    suspend fun clearCache() {
        context.gatewayDataStore.edit { it.remove(CACHED_API_URL) }
    }

    // MARK: - Private

    private suspend fun tryResolveViaENS(): String? {
        return try {
            // Fetch the ENS text record "gateway" from thevoteapp.eth
            // Using eth_call to PublicResolver — simplified via Cloudflare's JSON-RPC
            val cid = fetchENSTextRecord("thevoteapp.eth", "gateway") ?: return null
            fetchManifestFromCID(cid)
        } catch (e: Exception) {
            null
        }
    }

    private fun fetchENSTextRecord(name: String, key: String): String? {
        // Minimal ENS resolution: POST eth_call to Cloudflare ETH gateway
        // In production replace with a proper ENS client library
        val body = """
            {"jsonrpc":"2.0","id":1,"method":"eth_call","params":[{
              "to":"0x4976fb03C32e5B8cfe2b6cCB31c09Ba78EBaBa41",
              "data":"0x59d1d43c${nameHash(name)}${abiEncodeString(key)}"
            },"latest"]}
        """.trimIndent()

        val conn = URL(ENS_RESOLVER_URL).openConnection() as java.net.HttpURLConnection
        conn.apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/json")
            doOutput = true
            outputStream.write(body.toByteArray())
        }
        val response = conn.inputStream.bufferedReader().readText()
        val result = JSONObject(response).optString("result")
        return if (result.length > 2) decodeABIString(result) else null
    }

    private fun fetchManifestFromCID(cid: String): String? {
        for (gateway in IPFS_GATEWAYS) {
            try {
                val json = URL("$gateway$cid").readText()
                val manifest = JSONObject(json)
                val apiUrl = manifest.getJSONObject("endpoints").getString("api")
                if (apiUrl.startsWith("https://")) return apiUrl
            } catch (_: Exception) { }
        }
        return null
    }

    private fun getFallbackUrl(): String =
        "https://api.thevoteapp.org" // last-resort hardcoded URL

    // Minimal ENS namehash (RFC compliant)
    private fun nameHash(name: String): String {
        var node = ByteArray(32)
        name.split(".").reversed().forEach { label ->
            val labelHash = java.security.MessageDigest.getInstance("SHA-256")
                .digest(label.toByteArray())
            node = java.security.MessageDigest.getInstance("SHA-256")
                .digest(node + labelHash)
        }
        return node.joinToString("") { "%02x".format(it) }.padStart(64, '0')
    }

    private fun abiEncodeString(s: String): String {
        val bytes = s.toByteArray()
        val offset = "0000000000000000000000000000000000000000000000000000000000000020"
        val length = bytes.size.toString(16).padStart(64, '0')
        val data = bytes.joinToString("") { "%02x".format(it) }.padEnd(64, '0')
        return offset + length + data
    }

    private fun decodeABIString(hex: String): String? {
        return try {
            val clean = hex.removePrefix("0x")
            val offset = clean.substring(0, 64).toInt(16) * 2
            val len = clean.substring(offset, offset + 64).toInt(16)
            val strHex = clean.substring(offset + 64, offset + 64 + len * 2)
            strHex.chunked(2).map { it.toInt(16).toChar() }.joinToString("")
        } catch (_: Exception) { null }
    }
}

File: ios/VoteGlobal/Data/Gateway/GatewayResolver.swift
import CryptoKit
import Foundation

/**
 * GatewayResolver — discovers the live backend API URL without central DNS.
 *
 * Resolution order:
 *   1. UserDefaults cached URL from a previous successful resolution
 *   2. ENS text record on thevoteapp.eth (via Cloudflare ETH JSON-RPC)
 *   3. Hardcoded fallback URL baked into the app
 */
actor GatewayResolver {

    static let shared = GatewayResolver()

    private let cacheKey = "gateway_cached_api_url"
    private let ipfsGateways = [
        "https://ipfs.io/ipfs/",
        "https://cloudflare-ipfs.com/ipfs/",
        "https://gateway.pinata.cloud/ipfs/"
    ]
    private let ensRpcURL = URL(string: "https://cloudflare-eth.com/v1/mainnet")!
    private let fallbackURL = "https://api.thevoteapp.org"

    // MARK: - Public

    func resolveAPIURL() async -> String {
        // 1. Cache
        if let cached = UserDefaults.standard.string(forKey: cacheKey),
           !cached.isEmpty {
            return cached
        }
        // 2. ENS + IPFS manifest
        if let fromENS = await tryResolveViaENS() {
            await cacheURL(fromENS)
            return fromENS
        }
        // 3. Fallback
        return fallbackURL
    }

    func cacheURL(_ url: String) {
        UserDefaults.standard.set(url, forKey: cacheKey)
    }

    func clearCache() {
        UserDefaults.standard.removeObject(forKey: cacheKey)
    }

    // MARK: - Private

    private func tryResolveViaENS() async -> String? {
        guard let cid = await fetchENSTextRecord(name: "thevoteapp.eth", key: "gateway") else {
            return nil
        }
        return await fetchManifestFromCID(cid)
    }

    private func fetchENSTextRecord(name: String, key: String) async -> String? {
        let namehash = ensNameHash(name)
        let keyEncoded = abiEncodeString(key)
        // text(bytes32,string) selector: 0x59d1d43c
        let data = "0x59d1d43c" + namehash + keyEncoded

        let body: [String: Any] = [
            "jsonrpc": "2.0", "id": 1, "method": "eth_call",
            "params": [["to": "0x4976fb03C32e5B8cfe2b6cCB31c09Ba78EBaBa41", "data": data], "latest"]
        ]
        guard let bodyData = try? JSONSerialization.data(withJSONObject: body) else { return nil }

        var request = URLRequest(url: ensRpcURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = bodyData

        guard let (data, _) = try? await URLSession.shared.data(for: request),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let result = json["result"] as? String,
              result.count > 2
        else { return nil }

        return decodeABIString(result)
    }

    private func fetchManifestFromCID(_ cid: String) async -> String? {
        for gateway in ipfsGateways {
            guard let url = URL(string: "\(gateway)\(cid)") else { continue }
            guard let (data, _) = try? await URLSession.shared.data(from: url),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let endpoints = json["endpoints"] as? [String: Any],
                  let apiURL = endpoints["api"] as? String,
                  apiURL.hasPrefix("https://")
            else { continue }
            return apiURL
        }
        return nil
    }

    // MARK: - ENS helpers

    private func ensNameHash(_ name: String) -> String {
        var node = Data(repeating: 0, count: 32)
        let labels = name.split(separator: ".").map(String.init).reversed()
        for label in labels {
            let labelHash = SHA256.hash(data: Data(label.utf8))
            var combined = node
            combined.append(contentsOf: labelHash)
            node = Data(SHA256.hash(data: combined))
        }
        return node.map { String(format: "%02x", $0) }.joined().leftPad(toLength: 64, with: "0")
    }

    private func abiEncodeString(_ s: String) -> String {
        let bytes = Array(s.utf8)
        let offset = String(repeating: "0", count: 63) + "20"
        let length = String(bytes.count, radix: 16).leftPad(toLength: 64, with: "0")
        let data = bytes.map { String(format: "%02x", $0) }.joined()
            .padding(toLength: 64, withPad: "0", startingAt: 0)
        return offset + length + data
    }

    private func decodeABIString(_ hex: String) -> String? {
        let clean = hex.hasPrefix("0x") ? String(hex.dropFirst(2)) : hex
        guard clean.count >= 128 else { return nil }
        let offsetHex = clean.prefix(64)
        guard let offset = Int(offsetHex, radix: 16) else { return nil }
        let charOffset = offset * 2
        guard clean.count >= charOffset + 64 else { return nil }
        let lenHex = String(clean[clean.index(clean.startIndex, offsetBy: charOffset)...].prefix(64))
        guard let len = Int(lenHex, radix: 16), len > 0 else { return nil }
        let strStart = charOffset + 64
        guard clean.count >= strStart + len * 2 else { return nil }
        let strHex = String(clean[clean.index(clean.startIndex, offsetBy: strStart)...].prefix(len * 2))
        var result = ""
        var i = strHex.startIndex
        while i < strHex.endIndex {
            let next = strHex.index(i, offsetBy: 2)
            if let byte = UInt8(strHex[i..<next], radix: 16) {
                result.append(Character(UnicodeScalar(byte)))
            }
            i = next
        }
        return result
    }
}

private extension String {
    func leftPad(toLength length: Int, with pad: Character) -> String {
        guard count < length else { return self }
        return String(repeating: pad, count: length - count) + self
    }
}
